File: 백준/Gold/14499. 주사위 굴리기/주사위 굴리기.js
const fs = require("fs");
const input = fs.readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt").toString().trim().split("\n");

const [N, M, x, y, K] = input[0].split(" ").map(Number);
const graph = [];
for (let i = 1; i <= N; i++) {
    graph.push(input[i].split(" ").map(Number));
}
const cmds = input.at(-1).split(" ").map(Number);
const dice = {
    top : 0,
    bottom : 0,
    left : 0,
    right : 0,
    front : 0,
    back : 0,
};
const direction = {
    1 : [0, 1],   // 동쪽
    2 : [0, -1],  // 서쪽
    3 : [-1, 0],  // 북쪽
    4 : [1, 0]    // 남쪽
}

function roleDice(dir) {
    let tmp = 0;

    if (dir === 1) { // 동쪽
        tmp = dice.top;
        dice.top = dice.left;
        dice.left = dice.bottom;
        dice.bottom = dice.right;
        dice.right = tmp;

    } else if (dir === 2) { // 서쪽
        tmp = dice.top;
        dice.top = dice.right;
        dice.right = dice.bottom;
        dice.bottom = dice.left;
        dice.left = tmp;
    } else if (dir === 3) { // 북쪽
        tmp = dice.top;
        dice.top = dice.back;
        dice.back = dice.bottom;
        dice.bottom = dice.front;
        dice.front = tmp;
    } else if (dir === 4) { // 남쪽
        tmp = dice.top;
        dice.top = dice.front;
        dice.front = dice.bottom;
        dice.bottom = dice.back;
        dice.back = tmp;
    }
}

function sol() {
    const answer = [];
    let ci = x, cj = y;  // 초기 좌표 설정

    for (const cmd of cmds) {
        const dir = direction[cmd];
        const ni = ci + dir[0], nj = cj + dir[1];

        if (ni < 0 || ni >= N || nj < 0 || nj >= M) continue;  // 범위 밖으로 나가면 무시
        ci = ni, cj = nj;

        roleDice(cmd);
        answer.push(dice.top);

        if (graph[ci][cj] === 0) {
            graph[ci][cj] = dice.bottom;
        } else {
            dice.bottom = graph[ci][cj];
            graph[ci][cj] = 0;
        }
    }

    return answer.join("\n");
}

console.log(sol());

File: 백준/Gold/14499. 주사위 굴리기/주사위 굴리기.py
import sys
input = sys.stdin.readline

def solve():
    N, M, x, y, K = map(int, input().split())
    board = [list(map(int, input().split())) for _ in range(N)]
    cmds = list(map(int, input().split()))

    # 0: 위, 1: 아래, 2: 북, 3: 남, 4: 서, 5: 동
    dice = [0] * 6

    # 동, 서, 북, 남
    dx = [0, 0, 0, -1, 1]
    dy = [0, 1, -1, 0, 0]

    for cmd in cmds:
        nx = x + dx[cmd]
        ny = y + dy[cmd]

        if not (0 <= nx < N and 0 <= ny < M):
            continue

        t, b, n, s, w, e = dice

        if cmd == 1:      # 동
            dice = [w, e, n, s, b, t]
        elif cmd == 2:    # 서
            dice = [e, w, n, s, t, b]
        elif cmd == 3:    # 북
            dice = [s, n, t, b, w, e]
        else:             # 남
            dice = [n, s, b, t, w, e]

        if board[nx][ny] == 0:
            board[nx][ny] = dice[1]
        else:
            dice[1] = board[nx][ny]
            board[nx][ny] = 0

        print(dice[0])
        x, y = nx, ny

if __name__ == "__main__":
    solve()
